Allow shutdown during LoadMempool, dump only when necessary

jonasschnelli opened the pull request:
A try to address #9398.
Right now, `LoadMempool()` has now "interruption point". If one triggers a shutdown, It will always wait until `LoadMempool()` has been completed.
This PR does detect a shutdown during the `LoadMempool()` while also taking care of not-dumping the mempool during shutdown if the user has request it before the import was completed.

Includes also a GUI fix that ensure proper shutdowns (right now, `StartShutdown()` is not called when quitting the GUI).

ping @sipa 
Needs testing (maybe @instagibbs?)

jonasschnelli committed:
Allow shutdown during LoadMempool, dump only when necessary

paveljanik commented on src/validation.h:
```diff
@@ -557,7 +557,14 @@ static const unsigned int REJECT_CONFLICT = 0x102;
 void DumpMempool();
 
+/** Result states for loading the mempool from disk */
+enum LoadMempoolResult {
```
I think this is a bit overkill. `bool` is there already. And returning `true` only in case of successful `mempool.dat` load is enough. The only question is what to return in case of corrupted `mempool.dat`. Returning `true` instead of `false` makes sense here. And when `true` is returned, dump mempool at shutdown.

jonasschnelli replied:
IMO a simple `true|false` would not work. You need to distinct between...
* not yet loaded = don't dump during shutdown
* user_aborted (shutdown) = don't dump during shutdown
* failed = dump during shutdown
* successful = dump during shutdown

Returning `false` in case of a user-triggered shutdown could work, but then you would need to return `true` in case of an input error (somehow not good).

luke-jr replied:
If user aborts at T=5, and doesn't shut down the node, when it *does* finally shutdown at T=5000, you would want to dump it. This isn't possible, but the enum name doesn't make that clear. So a simply two-value enum with DoNotSave vs PleaseSave might be best?

paveljanik commented:
@luke-jr Two value enum aka bool? ;-)

MarcoFalke commented:
I might be missing something here but imo it is intentional to always dump the full mempool. In case the user feels it is taking too long (to load the mempool) on their hardware, they can switch to a smaller mempool with `-maxmempool=`?

paveljanik commented:
@MarcoFalke Imagine `mempool.dat` containing 300M of txs. User quits in the middle of loading it from the file. The mempool now contains ~150M of txs from the .dat file and some more txs from the network... What to do now?

luke-jr commented:
@paveljanik Using an enum would give it clear values which cannot be accidentally cast to/from a number.

jonasschnelli commented:
From a software design perspective, LoadMempool has three major states...
a) successfully loaded mempool
b) aborted, only partially loaded
c) error, something went wrong

The decision if we should dump during shutdown should probably not be part of `LoadMempool()`.
Only a) and c) should result in a dump during shutdown, otherwise, we overwrite a correct dump with a partially loaded mempool. But, if the mempool could not be loaded because of data-corruption (c), we want that to dump again.

sipa commented:
If you only have a partially loaded mempool file, but your node has been
running for a week after that, we should probably dump again...

jonasschnelli commented:
> If you only have a partially loaded mempool file, but your node has been
running for a week after that, we should probably dump again...

Yes. Partially loaded mempool is only possible with a user-triggered shutdown during `LoadMempool()`. This is why I named the stated `LoadMempoolResultUserAborted`.

paveljanik commented:
@jonasschnelli No, when the file is corrupted in the middle, you end up with partially loaded mempool.

jonasschnelli commented:
@paveljanik: Yes. Your right, I somehow though the mempool gets clear if a corruption was detected.
Simplified after recommendation from @sipa. Reverted from enum to bool.

instagibbs commented on src/init.cpp:
```diff
@@ -130,6 +130,7 @@ static const char* FEE_ESTIMATES_FILENAME="fee_estimates.dat";
 //
 
 std::atomic<bool> fRequestShutdown(false);
+std::atomic<bool> fDumpMempoolLater(false);
```
nit: `fDumpMempoolOnShutdown` seems more self-explanatory and precise.

instagibbs reviewed (commented):
Concept ACK

rebroad commented:
testing this. seems to be working well so far.

luke-jr commented:
utACK

sipa commented:
utACK ef479e19e9e9279c8e22cff50227ea0722f59df3

laanwj commented on src/qt/bitcoingui.cpp:
```diff
@@ -1105,6 +1105,15 @@ void BitcoinGUI::detectShutdown()
     }
 }
 
+void BitcoinGUI::shouldQuit()
```
Please don't make the qt shutdown sequence even more complicated :(

Wouldn't a better idea be to add StartShutdown here to app.requestShutdown, called from https://github.com/bitcoin/bitcoin/blob/master/src/qt/bitcoin.cpp#L691

jonasschnelli replied:
Argh. Yes. This is much simpler. Force push changed.
Request retests.

jonasschnelli committed:
[Qt] Do proper shutdown

TheBlueMatt commented:
utACK 325e400f9ba89c6f5501c6cedd4dee979a3912b2 I'd call this a pretty bad regression if we cant get this for 0.14.

MarcoFalke commented:
utACK 325e400

possientis commented:
building master branch commit **25720fc394e27a951bcad26095fb5a711bfacb8f** will revert asap.
